### PR TITLE
HDDS-10142. Add hidden command to set bucket encryption key to fix HDDS-7449

### PR DESCRIPTION
**Pull request by tanvipenumudy**
## What changes were proposed in this pull request?

Due to a known bug: '[HDDS-7449](https://issues.apache.org/jira/browse/HDDS-7449). Avoid overwriting bucket encryption properties when quota/replication config are set on an existing bucket', there can be consumers on a previous version of Ozone who could possibly run into this issue of losing bucket encryption properties when they (re)set quota or bucket replication configurations - which is a critical issue.

Through this patch, we can reset the BEK (bucket encryption key) of buckets affected by this bug through Ozone shell (`ozone sh bucket set-encryption-key -k <enckey> <vol>/<buck>`).

It does not change any other properties of the bucket or the properties of existing keys in the given bucket. 

- The properties of the older keys in the bucket remain as is.
- If any keys are added to the bucket after the bucket loses its encryption (before resetting the BEK), the keys added to the bucket remain unencrypted.
- The newer keys added to the bucket post resetting the BEK are encrypted using the BEK details provided (the `fileEncryptionInfo` object on the key-level gets generated as expected).

Under normal circumstances, we do not intend for users to (re)set bucket encryption post bucket creation. Therefore, the setter methods introduced, as well as the CLI class have been put behind a `@Deprecated` annotation.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10142

## How was this patch tested?

- Added integration tests
- Manually tested
- Green Git CI/CD checks: https://github.com/tanvipenumudy/ozone/actions/runs/7622391450/

### CODE REVIEW

**Comment by tanvipenumudy** [comment]
@adoroszlai, @fapifta could you please take a look? Thanks

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java
Comment (adoroszlai) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java:1002]: Please add javadoc and explain why it's deprecated, so readers don't have to dig up the PR.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:34]: I think it would be better to add this new command somewhere under `ozone debug`, to reduce exposure to users.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java
Comment (adoroszlai) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:36]: nit: let's omit `bucketkey` (or replace with `key`).  I know it's the same as in bucket properties, but we don't need to copy cumbersome options. ;)

**Review by adoroszlai — COMMENTED** [review]
Thanks @tanvipenumudy for the patch.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java
Comment (errose28) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:34]: Currently `ozone debug` is a read-only tool. HDDS-8100 has been on my backlog for a while now to add an `ozone repair` CLI that we can place these types of commands under. If required for this change I can begin working on it.

Comment (adoroszlai) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:34]: Thanks @errose28.  I think this comment from HDDS-8100:

> We should make sure the node being operated on is in the stopped state

does not apply to this "set bucket encryption" key command.  I guess we could still put it under `repair`, but must distinguish from offline tools in the future.

Comment (errose28) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:34]: That is true. The current proposal to enforce nodes being stopped while repaired would be to check their PID file, which happens in the shell script before invoking any Java code. Therefore making exceptions to that requirement may be hacky to implement, and the differing requirements for different commands may be confusing. Maybe making this subcommand hidden and leaving it under the bucket subcommand would be best.

Comment (adoroszlai) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:34]: https://picocli.info/#_hidden_subcommands

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java
Comment (tanvipenumudy) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java]: Got it, thank you @adoroszlai - made the changes!

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java
Comment (tanvipenumudy) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/bucket/SetEncryptionKey.java:62]: Thank you @adoroszlai and @errose28 for the review. 

I have retained the command under the `ozone sh bucket` set of commands but enabled the picocli `hidden` argument to `true` for the subcommand `set-encryption-key`.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java
Comment (tanvipenumudy) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/protocol/ClientProtocol.java:1016]: Thank you @adoroszlai, I have added a javadoc explaining why the functionality has been deprecated.

**Comment by adoroszlai** [comment]
Thanks @tanvipenumudy for updating the patch.  I'll let others take a look before merging it.

**Review by errose28 — APPROVED** [review]
LGTM thanks @tanvipenumudy

**Comment by adoroszlai** [comment]
Thanks @tanvipenumudy for the patch, @errose28 for the review.

**Pull request merged**
